### DISCUSSION TITLE
NextJS+ Socket.IO without custom server

### DISCUSSION THREAD

**Discussion by neilnahid** [vercel/next.js#17149] [Help]
Is it possible to integrate socket.io server without having to setup a custom server?

**Comment by jamesmosier** [comment]
Are you looking to just connect to a Socket.IO service? If so, there is an [old example](https://github.com/vercel/next.js/tree/442fbfaa4d4e8205cf10c7e5d27b8f51b7a4bdc6/examples/with-socket.io) that covers this.

**Reply by leerob** [reply]
You could also try using Pusher!

https://vercel.com/guides/deploying-pusher-channels-with-vercel

**Comment by shabbir-ahmed** [comment]
Apologize for this question! Is there any example of Next.js 9 with Socket.io? I have found one [article](https://www.valentinog.com/blog/socket-react/) but it is with a custom server but I want without a custom server like [this](https://nextjs.org/docs/api-routes/introduction)

It will be really appreciated if you help.

Thanks

**Reply by leerob** [reply]
I don't have an example but you might find some other good solutions here 😄 

https://vercel.com/docs/solutions/realtime

**Reply by insivika** [reply]
@shabbir-ahmed did you get this working without a custom server?

**Reply by kyds3k** [reply]
I'm curious as well, I'm trying to do something similar!

**Reply by j-mcgregor** [reply]
I was trying for a while myself, using NextJS with a combined server but I couldn't get it to work: making any change to the server restarted the whole app, which made it worse since I use Next Redux Wrapper and Redux Saga so it was really annoying. In the end I built a standalone Express server for Socket IO and it made all the difference

**Comment by simplenotezy** [comment]
While it seems it's difficult to run the actual Socket server from Next's API, perhaps it would make more sense to only emit events from their API, and then use a separate service or server to run the actual socket connection?

**Reply by rafde** [reply]
you can try https://stackoverflow.com/a/62547135
but I've not found a better solution

**Reply by saicharan-gevme** [reply]
> you can try https://stackoverflow.com/a/62547135 but I've not found a better solution

This method doesn't work anymore with Next.js 13. Does anyone know how to do it with Next.js 13?

**Reply by bvarga** [reply]
I'm creating a new HTTP server in [instrumentation.ts](https://nextjs.org/docs/app/building-your-application/optimizing/instrumentation) and using it for the WebSocket. Since it's running on a different port, I must also set up CORS headers in the socket server.

It would be nice to get a reference for the Next.js http server and initialize it in `instrumentation.ts`, but I haven't found a way yet.
